### DISCUSSION TITLE
webclient GET not blocking

### DISCUSSION THREAD

**Discussion by storchalberto** [arendst/Tasmota#21130] [Ideas]
I think it would be useful to have a non-blocking function equivalent to .GET until the response, so as not to block the system until the response arrives and to subsequently be able to verify the arrival of the response via get_size. It would therefore also be necessary to implement a function that allows you to recover the HTTP result when the get_size returns a value other than -1

**Comment by barbudor** [comment]
Just to be sure to understand your question
You are talking about Berry's WebClient ?
What you basically would like is an async mode with a call back ?

**Reply by storchalberto** [reply]
Yes. a call back will be the best, but if not possible, a non blocking get and a function (used inside a timer function) that allow to test the result

**Comment by s-hadinger** [comment]
There is a minimalistic async web client implemented for Matter bridge support. But it's primarily designed for Tasmota http, so very limited in features. 

https://github.com/arendst/Tasmota/blob/development/lib/libesp32/berry_matter/src/embedded/Matter_HTTP_async.be

**Reply by storchalberto** [reply]
I see, but i think it is not sufficient. i do not see the capability for using http

**Reply by s-hadinger** [reply]
What if you start explaining what you want to achieve?

**Reply by sfromis** [reply]
The "http" is in the name, meaning that it **is** http by definition. Did you check the examples at the bottom?

**Reply by storchalberto** [reply]
Here my example:
		  clt = webclient()
		  clt.begin("https://storchit.serversicuro.it/")
		  esito = clt.GET()
		  # here the machine is idle until the server responds
		  print("Result="+str(esito))
		  clt.close()
		  if esito ==200
		      print("Result OK")
		  else
	   	      print("Result ERROR")
		  end 
It runs ok but the berry machine is in hold during the wait for receive the answer.
I think that it will be very useful to avoid the bloching of the berry machine.

**Reply by s-hadinger** [reply]
I see. There is no support for https in this class. And no short term plans to implement it

**Reply by storchalberto** [reply]
Thanks, i am developing a system based on berry language that collect data and send to a server (in https as normal in actual internet) the result. the blocking of the webclient GET that i use to send to the host the collected data, is a big problem, because block the acquisition of the data input on sensors.

**Reply by s-hadinger** [reply]
Given that the underlying Arduino framework does not support non blocking IOs, I don’t have any solution for you

**Reply by storchalberto** [reply]
Ok i understand so: The async solution done in the matter bridge, is a product of your developement. the webclient class is part of the Arduino frameworks. Correct?

**Reply by sfromis** [reply]
Depending on your expectations for max time to send data, one strategy could be to offload the https processing to a second ESP32, where you can have local communication to that one. If you need to really minimize delay there, you could use [tcpclientasync](https://tasmota.github.io/docs/Berry/#tcpclientasync-class) or [udp](https://tasmota.github.io/docs/Berry/#udp-class).

**Reply by storchalberto** [reply]
OK, the use of a second esp32 solve the problem, but it needs to have 2 devices (and 2 different ip) on the wifi. and if i want to use Ethernet connection, it is necessary to have 4 ip address (2 wifi and 2 ethernet).I have seen that exist an arduino http async library, do you know about it?

**Reply by s-hadinger** [reply]
Yes the Matter async is a minimalistic reimplementation of http client in Berry on top of the non blocking tcp layer. The http client of arduino is blocking. And the TLS layer is not currently exposed to Berry, and I’m not sure its easy to expose

**Reply by storchalberto** [reply]
ok. I try to use 2 esp32 modules, with an udp interface. I have tried the udp protocol, hera the berry code for the server:
```
var conta=0
class udp_listener
  var u
  def init()
    conta=0
    self.u = udp()
    self.u.begin("",0)
    tasmota.add_driver(self)
  end
  def every_50ms()
    import string
    var packet = self.u.read()
    if packet != nil
        print("received")
	print(packet)
      #print(string.format(">>> Received packet ([%s]:%i): %s", self.u.remote_ip, self.u.remote_port, packet.tohex()), 2)
      #packet = self.u.read()
    else
	  conta=conta+1
	  if conta>20
	  	conta=0
	  	print("timeout")
	  end
    end
  end
end
d1 = udp_listener()
tasmota.add_driver(d1)
```
I use the windows program  Packet Sender to send an udp packet to the address of my device (192.168.1.77) but i do not see any received packet on the berry code. Suggestion?

**Reply by storchalberto** [reply]
I have also used the same code on the manual:
```
class udp_listener
  var u
  def init(ip, port)
    self.u = udp()
    print(self.u.begin_multicast(ip, port))
    tasmota.add_driver(self)
  end
  def every_50ms()
    import string
    var packet = self.u.read()
    while packet != nil
      tasmota.log(string.format(">>> Received packet ([%s]:%i): %s", self.u.remote_ip, self.u.remote_port, packet.tohex()), 2)
      packet = self.u.read()
    end
  end
end
```
but the result is the same: i do not receive the udp packets. I have also tried to use another ESP32 as sender:
   var u = udp()
    u.begin("",0)    
    u.send("192.168.1.77", 2000, bytes("414243"))   # send 'ABC' to 192.168.1.10:2000, source port is 2000
and the result is the same

**Reply by sfromis** [reply]
Did you try the Berry sample code for both receiver *and* sender?

One clear difference is that you use port 0 for a listener when the docs says:
> Listen on a specific `port` number, or set `0` to choose a random free port for **sending only**

About your 2nd post from just before I posted, you seem to be mixing up multicast vs not.

**Reply by storchalberto** [reply]
sorry: here the configuration:
ESP32 192.168.1.77
```
class udp_listener
  var u
  def init(ip, port)
    self.u = udp()
    print(self.u.begin("",0))
    tasmota.add_driver(self)
  end
  def every_50ms()
    import string
    var packet = self.u.read()
    while packet != nil
      tasmota.log(string.format(">>> Received packet ([%s]:%i): %s", self.u.remote_ip, self.u.remote_port, packet.tohex()), 2)
      packet = self.u.read()
    end
  end
end
```
ESP32 192.168.1.22
u = udp()
u.begin("", 2000)    # listen on all interfaces, port 2000
u.send("192.168.1.77", 2000, bytes("414243"))   # send 'ABC' to 192.168.1.77:2000, source port is 2000
And i not see anithing received on the server side.

**Reply by s-hadinger** [reply]
Doubling down on @sfromis remark, you need to choose a port to listen to. You can't "just send" UDP packets to an IP address, you need to specify a port number.

**Reply by s-hadinger** [reply]
You need to change from `self.u.begin("",0)` to `self.u.begin("",2000)`

**Reply by storchalberto** [reply]
Done but the same result:
ESP32 192.168.1.77
```
class udp_listener
  var u
  def init(ip, port)
    self.u = udp()
    print(self.u.begin("",2000))
    tasmota.add_driver(self)
  end
  def every_50ms()
    import string
    var packet = self.u.read()
    while packet != nil
      tasmota.log(string.format(">>> Received packet ([%s]:%i): %s", self.u.remote_ip, self.u.remote_port, packet.tohex()), 2)
      packet = self.u.read()
    end
  end
end
```
ESP32 192.168.1.22
```
u = udp()
u.begin("", 2000)    # listen on all interfaces, port 2000
u.send("192.168.1.77", 2000, bytes("414243"))   # send 'ABC' to 192.168.1.77:2000, source port is 2000
```
The server do not display received data

**Reply by s-hadinger** [reply]
Did you create an instance of `udp_listener`?

**Reply by storchalberto** [reply]
No. where is the documentation?

**Reply by sfromis** [reply]
Needing to create an instance after having defined a `class` is a **very** basic thing to doing programming with classes. Not doing it is like not running a program, but still expecting it to act.
https://github.com/berry-lang/berry/wiki/Reference

**Reply by storchalberto** [reply]
OK, here the code on the ESP32 192.168.1.77
```
class udp_listener
  var u
  def init(ip, port)
    self.u = udp()
    print(self.u.begin("",2000))
  end
  def every_50ms()
    import string
    var packet = self.u.read()
    while packet != nil
      tasmota.log(string.format(">>> Received packet ([%s]:%i): %s", self.u.remote_ip, self.u.remote_port, packet.tohex()), 2)
      packet = self.u.read()
    end
  end
end
d1 = udp_listener()
tasmota.add_driver(d1)
```
On the consolle i see: true.
here the code on ESP32 192.168.1.22
```
u = udp()
u.begin("", 2000)    # listen on all interfaces, port 2000
u.send("192.168.1.77", 2000, bytes("414243"))   # send 'ABC' to 192.168.1.77:2000, source port is 2000
```
I do not see on 192.168.1.77 the data received

**Reply by s-hadinger** [reply]
The example above works perfectly fine:
```
20:03:35.169 >>> Received packet ([192.168.2.210]:2000): 414243
```

I suspect that you are looking at the Berry console. Since you use `tasmota.log()`, it is not displayed in the Berry console but in the Tasmota console.

**Comment by TheHexaMaster** [comment]
https://github.com/arendst/Tasmota/pull/23836
